gadget/install/install.go

```diff
@@ -28,23 +28,39 @@ import (
 	"github.com/snapcore/snapd/boot"
 	"github.com/snapcore/snapd/gadget"
 	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil/disks"
 	"github.com/snapcore/snapd/secboot"
 	"github.com/snapcore/snapd/timings"
 )
 
-func deviceFromRole(lv *gadget.LaidOutVolume, role string) (device string, err error) {
+// deviceFromSystemSeedRole will identify an ondisk device node which is a disk
+// that has the SystemSeed role for the specified gadget volume on it.
+// It could be made to work for more generic roles, but some roles require
+// special handling like ubuntu-data, where on an encrypted system we really
+// need to look for a partition device node for
+// /dev/disk/by-partlabel/ubuntu-data-enc rather than /dev/.../ubuntu-data,
+// which is how this function currently operates and the logic to know whether
+// the device is an encrypted one or a normal one is not yet implemented, so we
+// artificially limit this function to only find the system seed role
+func deviceFromSystemSeedRole(lv *gadget.LaidOutVolume) (device string, err error) {
 	for _, vs := range lv.LaidOutStructure {
 		// XXX: this part of the finding maybe should be a
 		// method on gadget.*Volume
-		if vs.Role == role {
+		if vs.Role == gadget.SystemSeed {
 			device, err = gadget.FindDeviceForStructure(&vs)
 			if err != nil {
-				return "", fmt.Errorf("cannot find device for role %q: %v", role, err)
+				return "", fmt.Errorf("cannot find device for role %q: %v", gadget.SystemSeed, err)
```

Comment:
SInce the role is now hardcoded, maybe we can change the error message as well (also at the end of this function); BTW, I noticed that here you use `%q`, and there `%s`.